### ISSUE TITLE
Car: Different max speed for inner and outer city

### ISSUE THREAD

**Issue by karussell** [graphhopper/graphhopper#230]
Currenty max speed relies on the maxspeed tag in the city or village and often streets do not have such a tag. We need to grab the boundaries of the city and take this into account. 

This issue is related to #75 and #175

Possible improvements afterwards:
- [Route through Prague](https://graphhopper.com/maps/?point=50.1255%2C14.4588&point=50.0332%2C14.4003&locale=de&vehicle=car&weighting=fastest&elevation=true&layer=Omniscale) should be 20 minutes not 16
- [Laval, France](https://graphhopper.com/maps/?point=48.038496%2C-0.825005&point=48.096885%2C-0.70158&locale=de&vehicle=car&weighting=fastest&elevation=true&layer=Omniscale) should be 20 not 15 minutes and using one of the northern motorways not the southern (as first suggestion)
- [Mettmann Germany](https://graphhopper.com/maps/?point=51.255899%2C6.939411&point=51.250527%2C7.014427&locale=de&vehicle=car&weighting=fastest&elevation=true&layer=Omniscale) should avoid the inner city and slightly slower (not 8min but more like 10min)
- [Bayreuth, Germany](https://graphhopper.com/maps/?point=49.956963%2C11.582508&point=49.936085%2C11.549892&locale=de&vehicle=car&weighting=fastest&elevation=true&layer=Omniscale) should use the nordring (as first suggestion) and avoid the inner city, also not 5min but more like 10
- [Paris](https://graphhopper.com/maps/?point=48.924304%2C2.603073&point=48.826983%2C2.243614&locale=de&vehicle=car&weighting=fastest&elevation=true&layer=Omniscale) oh, not through Paris! And not 33min but more like 40-45min (at least)
- [London](https://graphhopper.com/maps/?point=51.556369%2C-0.155525&point=51.458499%2C-0.09613&locale=de&vehicle=car&weighting=fastest&elevation=true&layer=Omniscale) Time is 20 but should be more like 40min (but here going through the city is similar slow ;))
- [London](https://graphhopper.com/maps/?point=Cricklewood%2C%20NW2%202DQ%2C%20London%2C%20United%20Kingdom&point=51.539983%2C-0.178871&locale=de&vehicle=car&weighting=fastest&elevation=true&layer=Omniscale) should be more like 10-15 minutes not 6
- why is the time difference so big [here](https://graphhopper.com/maps/?point=49.930119%2C11.481228&point=49.937411%2C11.419258&point=49.921057%2C11.383295&locale=de&vehicle=car&weighting=fastest&elevation=true&layer=Omniscale)? (9min vs. 6min)

(These numbers and 'real routes' are from mostly customers&user reports and own experiences)

**Comment by jansoe** [comment]
I currently work on this, but I guess city boundaries are not an aedequate solution.
One might could use ´´´traffic_sign = city_limit´´´, but it seems that this tag is frequently not set.
For this reason I propose right now, to use the landuse tag and take ´´´residential´´´ and ´´´commercial´´´ areas as areas with reduced speed limit.

**Comment by loffenauer** [comment]
Hi-
we have worked on this for our MindYourSpeed app (see
http://www.volzinnovation.de/MindYourSpeed/). The approach that we took was
a preprocessing step at the OSM data level. We split streets and tagged
them accordingly based on their containment within a city boundary.
It is a time consuming task.
The automated assignment of speed limits took into account the various
legal requirements of permissible max speeds per way type based on the
legal requirements in each country
(we only worked with data for Switzerland, Austria, France, Netherlands,
Germany).


The preprocessing took about 1 day per 200 MB of OSM data...

Best regards,


Raphael


On 18 February 2015 at 13:59, jansoe <notifications@github.com> wrote:

> I currently work on this, but I guess city boundaries are not an aedequate
> solution.
> One might could use ´´´traffic_sign = city_limit´´´, but it seems that
> this tag is frequently not set.
> For this reason I propose right now, to use the landuse tag and take
> ´´´residential´´´ and ´´´commercial´´´ areas as areas with reduced speed
> limit.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/graphhopper/graphhopper/issues/230#issuecomment-74859731>
> .
>



-- 

--
"All ideas are good ones as long as you don't think about them."
( Michael Cembalest )

--
Prof. Dr. Raphael Volz
Vorsitzender des Aufsichtsrats
chairman of the board

Volz Innovation GmbH
Kirchhaldenweg 10, 76597 Loffenau, Germany

Web: www.volzinnovation.com
Phone: +49 7083 9328338
Fax: +49 7083 9328337

Company Seat / Sitz der Gesellschaft: Loffenau
Company Register / Handelsregisternummer: HRB 704418, Amtsgericht Mannheim
Managing Director / Geschäftsführer: Dr. Sara Brockmans
Chairperson of the Supervisory Board / Vorsitzender des Aufsichtsrats:
Prof. Dr. Raphael Volz

**Comment by jansoe** [comment]
what do you use as boundary? I think administrative boundaries are too loose (e.g. http://www.openstreetmap.org/relation/66527#map=13/51.2105/7.0547)? 
Is it also landuse=residential (http://www.openstreetmap.org/relation/1537220#map=14/51.1898/7.0176)?

**Comment by loffenauer** [comment]
Hi,
we used landuse=residential
http://taginfo.openstreetmap.org/keys/landuse#values
as well as other landuse way tags (industrial, commercial, retail, quarry,
village_green, etc.), but they get more and more sparse.
So residential, industrial, commercial and retail capture the most

We only assigned values, if there was no maxspeed tag originally in the
data. We considered both relations and ways and created new way ids for the
splitted ways.

Best regards,

Raphael


On 18 February 2015 at 15:17, jansoe <notifications@github.com> wrote:

> what do you use as boundary? I think administrative boundaries are too
> loose (e.g.
> http://www.openstreetmap.org/relation/66527#map=13/51.2105/7.0547)?
> Is it also landuse=residential (
> http://www.openstreetmap.org/relation/1537220#map=14/51.1898/7.0176)?
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/graphhopper/graphhopper/issues/230#issuecomment-74870043>
> .
>



-- 

--
"All ideas are good ones as long as you don't think about them."
( Michael Cembalest )

--
Prof. Dr. Raphael Volz
Vorsitzender des Aufsichtsrats
chairman of the board

Volz Innovation GmbH
Kirchhaldenweg 10, 76597 Loffenau, Germany

Web: www.volzinnovation.com
Phone: +49 7083 9328338
Fax: +49 7083 9328337

Company Seat / Sitz der Gesellschaft: Loffenau
Company Register / Handelsregisternummer: HRB 704418, Amtsgericht Mannheim
Managing Director / Geschäftsführer: Dr. Sara Brockmans
Chairperson of the Supervisory Board / Vorsitzender des Aufsichtsrats:
Prof. Dr. Raphael Volz

**Comment by jansoe** [comment]
Thanx Raphael for your input.
My approach is very similar, but to speed computation up I will assign speed limits not exactly on point but within a pixel raster. Do you have an estimate how well landuse boundarys correspond to ```city_limit``` areas?

**Comment by NopMap** [comment]
Some comments on this:
It would be good to develop a general solution. Assigning a speed value within city boundaries is only one application. Other (currently unresolved) use cases are:
- marking all ways inside an area as unaccessible, e.g. in a military restricted area
- applying rules to all ways inside a natural reserve area
- applying legal restrictions to all ways inside woods (there are such laws in some German states)

Putting this into a preprocessing step feels like a good idea to me. So it does not weigh down the regular import and may be executed on a different machine. I assume the memory requirements are stiff. Are your solutions based on a database?

A spontaneous idea for optimization is merging the various polygons into a single area first, then matching the ways against the smaller number of merged areas.

bye, Nop

**Comment by karussell** [comment]
Yes, other usecase could be prefer ways for bikes in a forest etc. But a generic solution will follow. This is tricky on its own, so making it generic from the start is overkill. As always: step by step :)

> I assume the memory requirements are stiff

For worldwide we expect roughly 10% more RAM.

> A spontaneous idea for optimization is merging the various polygons into a single area first

A similar approach will be used (a 'simple' grid assigning boolean values)

> Are your solutions based on a database?

No. GraphHopper will not depend on an external thing to administrate. An internal java based database might be used for the import though. See #220

**Comment by jansoe** [comment]
> It would be good to develop a general solution
it's right now a simple solution, but with future generalization in mind. i hope it's ready beginning of next week for evaluation.

**Comment by jansoe** [comment]
> Currenty max speed relies on the maxspeed tag in the city or village and often streets do not have such a tag

My investigations suggest, that it is not missing maxspeed tags which cause counter intuitive routes, but rather the assumption that both small and larger roads with the same maxspeed also result in the same average speed, neglecting prioritised traffic flow on larger roads. 
Therefore in this [branch](https://github.com/jansoe/graphhopper/tree/CitySpeedMerge) I apply the following heuristic to estimate average travel velocity: the travel speed of ways within defined areas (e.g. residential, commercial, retail) is reduced by different factors on primary, secondary, tertiary roads.

In my opininon this leads to more intuitive roads

old:
![oldspeed](https://cloud.githubusercontent.com/assets/903347/6789920/3a0c06de-d1a4-11e4-8e83-e9580f93ac52.png)
new:
![newspeed](https://cloud.githubusercontent.com/assets/903347/6789926/42d0bd00-d1a4-11e4-8223-e49f26f7c194.png)

old:
![oldspeedouter](https://cloud.githubusercontent.com/assets/903347/6789946/627cf628-d1a4-11e4-9fec-ee428c46673f.png)
new:
![newspeedouter](https://cloud.githubusercontent.com/assets/903347/6789936/4ced799a-d1a4-11e4-8733-11fb8ffb992b.png)

**Comment by karussell** [comment]
You mean that these improvements are possible without the difference of inner and outer city?

**Comment by jansoe** [comment]
No, I mean it is not sufficient to adjust values where the maxspeed tag is missing, but also to adjust maxspeed values "within" a city by a factor depending on the [highway level](https://github.com/jansoe/graphhopper/blob/CitySpeedMerge/core/src/main/java/com/graphhopper/routing/util/CarFlagEncoder.java#L110)

**Comment by karussell** [comment]
Thanks for clarification & nice! I'll play with it tomorrow!

**Comment by jansoe** [comment]
All in all this increases travel time by about 10%.
In my opinion this still does not reflect true travel times, but in a second step one might take turn cost and traffic lights into consideration.
(see #334, #75)

I think an optimal solution would take real world travel times into account and would adjust the heuristic to reflect them. But for this one would need an extensive dataset of such travel data. Anyone aware of this?

**Comment by jansoe** [comment]
This [branch](https://github.com/jansoe/graphhopper/tree/CitySpeedMerge) now only assigns speed values to streets within citys where the maxspeed tag is missing. But this has only a minor effect on traveltimes. Also, the approach to use ``` landuse = residential | industrial | retail ``` to detect city regions, seems to be a bit problematic, because [there are citys (e.g. madrid) where large areas are not tagged](http://overpass-turbo.eu/s/8SM). (Although for germany it seems OK).

In general this code allows to apply region specific speed limits. So one might could reuse this code to implement country specific speed limits (#175) or for other cases.

**Comment by ratrun** [comment]
@jansoe: Your above referenced branch tree/CitySpeedMerge returns a now "404 Not found". I wanted to take a look on what you did in this branch. Could you provide this work again?

**Comment by karussell** [comment]
@ratrun @jansoe the branch would be helpful for me too :)

**Comment by jansoe** [comment]
@ratrun  @karussell 
sorry for the delayed reaction time, but now the branch is again online. don't hesitate to ask questions, if anything seems unclear

**Comment by dteirney** [comment]
@karussell apologies for nudging an old issue. Is there some guidance or direction (perhaps elsewhere) that you could share for options to help with routing and the associated duration calculations within urban centers? We're finding that we get a lot of routes traversing through urban centers with lots of traffic lights where it's very unlikely that the maxspeed on those ways will be accurate for an average speed approximation when compared to the maxspeed for nearby motorway routing options.

**Comment by karussell** [comment]
See also the recent discussion in #2395. This should be a foundation to improve the ETAs and avoiding city centers.
